rfcs/20220217-visual_regression_testing.md

```diff
@@ -0,0 +1,226 @@
+ *
+ * We can also automatically sync the local playwright version
+ * with the docker image version. This ensures that the browser
+ * versions expected by the `playwright` binary installed locally
+ * matches the versions pre-installed in the container.
+ */
+const { spawnSync } = require('child_process')
+const fs = require('fs')
+const path = require('path')
+const yaml = require('yaml')
+
+const pnpmLock = yaml.parse(
+  String(fs.readFileSync(path.resolve(process.cwd(), 'pnpm-lock.yaml')))
+)
+
+const playwrightVersion = pnpmLock.devDependencies['@playwright/test']
+
+const args = [
+  'run',
+  '--rm',
+  '-it',
+  '--mount',
+  `type=bind,source=${process.cwd()},target=/src`,
+  '--add-host=host.docker.internal:host-gateway', // This is necessary to make `host.docker.internal` work on Linux. Does it break the command for other OSs?
+  `mcr.microsoft.com/playwright:v${playwrightVersion}-focal`,
+  '/src/bin/visual-regression.sh',
+]
+
+if (process.argv.includes('-u')) {
+  args.push('-u')
+}
+
+spawnSync('docker', args, {
+  stdio: 'inherit',
+})
+2. Create a `./bin/visual-regression.sh` file containing the following (don't forget to `chmod +x` it):
+```sh
+#!/bin/bash
+# This script is meant to run inside the playwright docker container, hence the assumed `/src` directory
+cd /src
+
+# Use npm to avoid having to install pnpm inside the container, it doesn't matter in this case
+npm run test:visual-regression:local -- $1
+```
+3. Add two new package.json scripts:
+```json
+"test:visual-regression": "node ./bin/visual-regression.sh",
+"test:visual-regression:local": "playwright test -c ./test/visual-regression"
+```
+
+All of this together should successfully allow us to run the snapshot tests inside a docker container.
+
+For local debugging purposes, `pnpm test:visual-regression:local --debug` can be used to run the snapshot tests headed in a browser on the host OS. This is useful for debugging the playwright scripts before they actually take a screenshot. Just keep in mind that the final screenshots submitting in a PR need to come from the docker container.
+
+## Recommended tooling
+
+* Continue using Playwright to manipulate the browser and use it's `screenshot` function to generate screenshots.
+* Use `jest-image-snapshot` to save and diff screenshots.
+* Run snapshot tests inside a docker container to avoid cross-platform rendering differences.
+* Write a GitHub action to be run [once daily using the `on.schedule` feature](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onschedule) to automatically remove snapshots from the repository history older than a month old.
+* Use GitHub's PR review UI to review visual diffs of snapshots.
+
+If agreeable with the team, I'd also like to take this opportunity to implement TypeScript linting and recommend that we switch to using TypeScript to write Playwright tests. Playwright supports it natively.
```

Comment:
Considering that we've already added typescript parser to ESLint, we can omit this sentence (and replace 'Typescript setup' with 'Conversion of the existing tests to TS'.

Reply:
Good catch, removed those sections from the RFC in https://github.com/WordPress/openverse/pull/165/commits/f411fda12a196fb5e18b9559bfd923805bbf4b6e